qutip/solver/mcsolve.py

```diff
@@ -0,0 +1,311 @@
+
+    def prob_func(self, state):
+        if self.rhs.issuper:
+            return _data.norm.trace(unstack_columns(state))
+        return _data.norm.l2(state)**2
+
+    def norm_func(self, state):
+        if self.rhs.issuper:
+            return _data.norm.trace(unstack_columns(state))
+        return _data.norm.l2(state)
+
+    def _step(self, t, copy=True):
+        t_old, y_old = self._integrator.get_state(copy=False)
+        norm_old = self.prob_func(y_old)
+        while t_old < t:
+            t_step, state = self._integrator.mcstep(t, copy=False)
+            norm = self.prob_func(state)
+            if norm <= self.target_norm:
+                self._do_collapse(norm_old, norm, t_old)
+                t_old, y_old = self._integrator.get_state(copy=False)
+                norm_old = 1.
+            else:
+                t_old, y_old = t_step, state
+                norm_old = norm
+
+        return t_old, _data.mul(y_old, 1 / self.norm_func(y_old))
+
+    def _do_collapse(self, norm_old, norm, t_prev):
+        """Find and apply a collapse."""
+        t_final, _ = self._integrator.get_state()
+        tries = 0
+        while tries < self.options.mcsolve['norm_steps']:
+            tries += 1
+            if (t_final - t_prev) < self.options.mcsolve['norm_t_tol']:
+                t_guess = t_final
+                _, state = self._integrator.get_state()
+                break
+            t_guess = (
+                t_prev
+                + ((t_final - t_prev)
+                   * np.log(norm_old / self.target_norm)
+                   / np.log(norm_old / norm))
+            )
+            if (t_guess - t_prev) < self.options.mcsolve['norm_t_tol']:
+                t_guess = t_prev + self.options.mcsolve['norm_t_tol']
+            _, state = self._integrator.mcstep(t_guess, copy=False)
+            norm2_guess = self.prob_func(state)
+            if (
+                np.abs(self.target_norm - norm2_guess) <
+                self.options.mcsolve['norm_tol'] * self.target_norm
+            ):
+                break
+            elif (norm2_guess < self.target_norm):
+                # t_guess is still > t_jump
+                t_final = t_guess
+                norm = norm2_guess
+            else:
+                # t_guess < t_jump
+                t_prev = t_guess
+                norm_old = norm2_guess
+
+        if tries >= self.options.mcsolve['norm_steps']:
+            raise Exception("Norm tolerance not reached. " +
+                            "Increase accuracy of ODE solver or " +
+                            "SolverOptions.mcsolve['norm_steps'].")
+
+        # t_guess, state is at the collapse
+        probs = np.zeros(len(self._n_ops))
+        for i, n_op in enumerate(self._n_ops):
+            probs[i] = n_op.expect_data(t_guess, state).real
+        probs = np.cumsum(probs)
+        which = np.searchsorted(probs, probs[-1] * self.generator.random())
```

Comment:
A minor thing to consider: the current documentation for [montecarlo](https://qutip.org/docs/latest/guide/dynamics/dynamics-monte.html) says that the chosen collapse operator is based on "r" which I think here would be target_tol and not a new random number. Not sure if this has really a significant effect on the result (I think it does not) but I would suggest to use tartget_norm as the documentation says. What was using the previous method?

Reply:
The documentation is wrong, it needs to be another random number. It was fixed in  #1677.